### ISSUE TITLE
No karma entry point needed?

### ISSUE THREAD

**Issue by BrainCrumbz** [wkwiatek/angular2-webpack#1]
Hello there,
thanks for sharing your take on Angular2 + webpack setup!

We noticed a karma entry point source file in other starter projects (e.g. [AngularClass](https://github.com/AngularClass/angular2-webpack-starter/blob/master/spec-bundle.js) and [preboot](https://github.com/preboot/angular2-webpack/blob/master/karma-shim.js)), where usually:
1. some `Error.stackTraceLimit` is set to `Infinity`
2. a number of modules are required, 
3. some `context` is built out of `require.context()` from spec file patterns, and then processed in some funky way
4. Select some `BrowserDomAdapter` so to fix a known issue

Such an entry point seems to be missing in this project. I can see the required modules are pulled in in [karma.conf.js files section](https://github.com/wkwiatek/angular2-webpack2/blob/master/karma.conf.js), so #2 should be mapped. What about the remaining points? Did you have any apparent issue?

In other words, does one actually need that entry point or not? Thanks

**Comment by BrainCrumbz** [comment]
Ah, I think you have #4 as well, only in App spec [here](https://github.com/wkwiatek/angular2-webpack2/blob/960e50b2c0a8eac17d3f1552f9da5a0d0496d667/src/app.spec.ts#L10)

**Comment by wkwiatek** [comment]
Thanks for sharing your thoughts!

Yes - I've seen these ones too. Both of them are really great but are trying to cover every single case (which I didn't need).

So to answer your points:
1. You can take a look here to see why it can be helpful: http://stackoverflow.com/questions/7697038/more-than-10-lines-in-a-node-js-stack-error. It's not here because simply I haven't run yet into troubles like these (but i think it is worth to add it just in case).
2. Required modules are directly in `karma.conf.js` and are bundled using webpack config which is also directly in this file.
3. It is to require files using some wildcards. The setup done there I've put into `files` in `karma.conf.js`
4. As you mentioned - it's in test itself however I think it would be better to put it into separated file because currently you have to add it into each test. I'll do it in some spare time.

So to wrap it up. It's kinda different way to achieve the same thing basically. I just wanted this setup to be really minimal for both e2e and unit tests.

**Comment by BrainCrumbz** [comment]
Thanks a lot for clarifying! 

We're trying to understand if we can avoid the test shim altogether, and resolve everything with `karma.conf.js`, or above all avoid _strange_ parts like `testContext.keys().forEach(testContext)` and keep things as clear as possible. 

So, trying to understand what's the cleanest way to hit the target:
1. infinite stack trace limit: if needed, this can only go in test shim, it cannot be moved to `karma.conf.js`. Its aim is understandable with a comment
2. required modules: can be listed in `karma.conf.js`, instead of being required in test shim for the sake of it
3. test spec files: can be listed in `karma.conf.js`, instead of being required and processed in test shim in a somehow strange way
4. BrowserDomAdapter selection: this can only go in test shim, it cannot be moved to `karma.conf.js`. No point in repeating in all spec files. Its aim is understandable with a comment

**Comment by wkwiatek** [comment]
I think the way you summarised it is the one which is as simple as it can be. 

If you have any suggestions to what could be changed here in the starter please feel free to share.

**Comment by BrainCrumbz** [comment]
We tried those 4 actions and it seems to work great. So, nr.2 and nr.3 are directly taken from you starter, nothing more to add there than _"thank you!"_. I'd suggest nr.1 and nr.4, so have a very slim test shim to the purpose, this is ours:

``` javascript
// karma-entry.js
/*
 * This is the entry file for karma tests, with some test environment setup code 
 * needed across all tests running in the browser.
 */

// Turn on full stack traces in errors to help debugging
Error.stackTraceLimit = Infinity;

// Somewhere in the test setup, select BrowserDomAdapter to run tests in the browser
// see https://github.com/AngularClass/angular2-webpack-starter/issues/124
var testing = require('angular2/testing');
var browser = require('angular2/platform/testing/browser');

testing.setBaseTestProviders(
  browser.TEST_BROWSER_PLATFORM_PROVIDERS,
  browser.TEST_BROWSER_APPLICATION_PROVIDERS);
```

As mentioned, `karma.conf.js` has to reference both required modules and test spec, plus the test shim itself:

``` javascript
// karma.conf.js
// ...
    // list of files/patterns to load in the browser, order is important
    files: [
      // required third-party modules
      { pattern: 'node_modules/es6-shim/es6-shim.js', included: true, watched: false }, 
      { pattern: 'node_modules/es6-promise/dist/es6-promise.js', included: true, watched: false }, 
      { pattern: 'node_modules/reflect-metadata/Reflect.js', included: true, watched: false }, 
      { pattern: 'node_modules/zone.js/dist/zone-microtask.js', included: true, watched: false }, 
      { pattern: 'node_modules/zone.js/dist/long-stack-trace-zone.js', included: true, watched: false }, 
      // shim entry point to build the test environment
      { pattern: path\to\your\karma-entry-point.js, included: true, watched: false }, 
      // the actual test spec
      { pattern: pattern\selecting\your\test\spec\files\*, included: true, watched: true },
    ],
```

We also had to keep an eye on pre-processing both the test shim (for the 2 `require` in there, I think) and of course the spec files:

``` javascript
// karma.conf.js
// ...
  // Preprocess matching files before serving them to the browser
  // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
  preprocessors: {

    'path\to\your\karma-entry-point.js': [
      'webpack', // compile TS and ES6 files  
      'sourcemap', // generate source maps
    ],

    'pattern\selecting\your\test\spec\files\*': [
      //'coverage', // generate coverage for those - ISSUE: this breaks build
      'webpack', // compile TS and ES6 files  
      'sourcemap', // generate source maps
    ],

  },
```

As you can see, we still have an issue when adding code coverage preprocessing ... not sure if that is really needed or not to get coverage results.
HTH

**Comment by BrainCrumbz** [comment]
Spoke too soon: we had to move back collecting spec files from Karma config to test entry point. The reason being that when running Karma tests in a _headed_ browser (e.g. not PhantomJS) and looking at them in developer toolbar (e.g. to put a breakpoint), Chrome is not able to pick original `.ts` files, apparently because is not able to pick the corresponding source maps, so you're left with browsing long, bundled JS files.

We double checked our setup and can confirm that `karma-sourcemap-loader` is included and applied. Several articles (also for React, which transpiles anyway) all mention to explicitly `require` spec files from test shim, so that webpack can process them to transpile and generate source maps. Apparently, it seems that including them in `karma.conf.js` is not enough to get source maps as well, which sounds somehow reasonable.

@wkwiatek are you able to step through original TypeScript spec code with your current setup?

BTW we managed to understand (and better comment) the purpose of that obscure `testContext.keys().forEach(testContext)`.

**Comment by wkwiatek** [comment]
Sourcemaps to typescript files are not working in this project. I haven't had time to figure it out why. I'm pretty sure it's not the point of having single entry point for webpack files. 

The reason you see something like `testContext.keys().forEach(testContext)` you can find in `karma-webpack` docs: https://github.com/webpack/karma-webpack#alternative-usage.

If you find any way to make it work feel free to submit PR or just share here.

**Comment by BrainCrumbz** [comment]
Thanks for pointing those docs: that (and others) helped us better understand the meaning. We did find a way to have sourcemaps working with typescript, although we're not totally satisfied with it.

Just to be clear: we do have sourcemaps working with a single test shim entry point (i.e. the _alternative usage_ mentioned in those docs). What we did not fulfill yet is to have sourcemaps working with spec files loaded in Karma conf and not in test shim (i.e. the _normal_ usage described in https://github.com/webpack/karma-webpack#usage and in your project as well).

### Step 1

In order to have typescript test code browsable & debuggable from browser developer tools, we had to require spec source files from test shim. We left spec source files listed in karma config only to be watched over:

``` javascript
// karma.conf.js
//...
    files: [
      // required third-party modules
      { pattern: 'node_modules/es6-shim/es6-shim.js', included: true, watched: false }, 
      // ...
      // shim entry point, to build test environment and run all spec files
      { pattern: paths.testEntry, included: true, watched: false }, 
      // the actual test spec, only to be monitored for re-runs
      { pattern: patterns.testSources, included: false, watched: true },
    ],
// ...
   config['preprocessors'][paths.testEntry] = [
    'webpack', // compile TS and ES6 files  
    'sourcemap', // generate source maps
  ];
// ...
```

``` javascript
// webpack.config.test.js
// ...
    loaders: [
      { 
        test: /\.ts$/, 
        loaders: ['ts'],
        include: [
          '/path/to/source/code',
        ], 
        exclude: [
          paths.nodeModules, // skip all node modules
          paths.typings, // skip all type definitions
          paths.dist, // skip output
          /\.(e2e|async)\.ts$/, // skip end-to-end test and async TS files 
        ],
      },
      // ...
    ],
```

### Step 2

That worked for typescript spec files, but then we found out that typescript _main_ source files were browsable, but all mangled up with some `__cov`-prefixed variables and functions. Up to now, the only way we found to have clear typescript code is to disable istanbul instrumentation post-loader. So basically we commented out the following block:

``` javascript
// webpack.config.test.js
// ...
    postLoaders: [
      // instrument only code that isn't test or third-party 
      // delay coverage until after tests are run, fixing transpiled source coverage error
      istanbul: 
      {
        test: /\.(js|ts)$/,
        loaders: ['istanbul-instrumenter'],
        include: [
          '/path/to/source/code',
        ], 
        exclude: [
          /\.(e2e|spec)\.ts$/, // skip all test files
          paths.nodeModules, // skip all node modules
          paths.typings, // skip all type definitions
          paths.dist, // skip output
        ],
      },
    ],
```
